### ISSUE TITLE
Adding primary keys to tables

### ISSUE THREAD

**Issue by isedwards** [opencdms/climsoft-database#2]
@smachua could you look through the Climsoft database for tables that don't currently have primary keys and suggest which fields can make up the primary keys?  In many cases these already have unique indices.

We can then create a script with `ALTER TABLE table_name ADD primary key (field1, field2, ...)` commands as part of the next database update.

**Comment by smachua** [comment]
Ok @isedwards I will look through.

**Comment by isedwards** [comment]
Thank you Samuel

**Comment by isedwards** [comment]
Noting #1: `DROP TABLE IF EXISTS abc, aws1, aws_elements, aws_stations, aws_test, ss, testing_aws, testing_aws1;`

The table below lists every table in the Climsoft 4.1.1 schema that currently does not explicitly define a primary key. The "Candidate Primary Key" gives the current `UNIQUE INDEX` in cases where this could be explicitly made into a composite primary key.

For completeness, the third column compares this analysis to the attached primary keys summary:
 [Climsoft.Data.Model.Tables.and.Keys.pdf](https://github.com/climsoft/climsoft-database/files/7006467/Climsoft.Data.Model.Tables.and.Keys.pdf) (lists 24 tables)

<table>
<tr><th>Table name</th><th>Candidate Primary Key</th><th>Comparison</th>
<tr><td>    aws_lsi                       </td><td> UNIQUE INDEX `identification` (`Cols`)</td><td></td></tr>
<tr><td>    aws_lsi1                      </td><td> UNIQUE INDEX `identification` (`Cols`)</td><td></td></tr>
<tr><td>    aws_malawi12                  </td><td> UNIQUE INDEX `identification` (`No`)</td><td></td></tr>
<tr><td>    aws_rema1                     </td><td> UNIQUE INDEX `identification` (`Cols`)</td><td></td></tr>
<tr><td>    aws_rwanda1                   </td><td> # no unique candidate: `Cols` INT(11) NULL DEFAULT '0'</td><td></td></tr>
<tr><td>    aws_rwanda_rain               </td><td> UNIQUE INDEX `identification` (`Cols`)</td><td></td></tr>
<tr><td>    aws_toa5_bw1                  </td><td> UNIQUE INDEX `identification` (`Element_Abbreviation`) # `Cols` not suitable</td><td></td></tr>
<tr><td>    bufr_crex_data                </td><td> # no unique candidate: `nos` INT(11) NULL DEFAULT '0',</td><td></td></tr>
<tr><td>    bufr_crex_master              </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    ccitt                         </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    faultresolution               </td><td> UNIQUE INDEX `solution` (`resolvedDatetime`, `associatedWith`)</td><td>✓</td></tr>
<tr><td>    featuregeographicalposition   </td><td> UNIQUE INDEX `FK_mysql_climsoft_db_v4_synopfeatureFeatureGeographicalPosition` (`belongsTo`, `observedOn`)</td><td>✓</td></tr>
<tr><td>    instrumentinspection          </td><td> UNIQUE INDEX `inspection` (`performedOn`, `inspectionDatetime`)</td><td>✓</td></tr>
<tr><td>    observationfinal              </td><td> UNIQUE INDEX `obsFinalIdentification` (`recordedFrom`, `describedBy`, `obsDatetime`)</td><td>add: obsLevel</td></tr> 
<td>    observationinitial            </td><td> UNIQUE INDEX `obsInitialIdentification` (`recordedFrom`, `describedBy`, `obsDatetime`, `qcStatus`, `acquisitionType`)</td><td>add: obsLevel</td></tr> 
<tr><td>    observationschedule           </td><td> UNIQUE INDEX `scheduleIdentification` (`classifiedInto`, `startTime`, `endTime`)</td><td>✓</td></tr>
<tr><td>    paperarchive                  </td><td> UNIQUE INDEX `paper_archive_identification` (`belongsTo`, `formDatetime`, `classifiedInto`)</td><td>✓</td></tr>
<tr><td>    physicalfeature               </td><td> UNIQUE INDEX `featureIdentification` (`associatedWith`, `beginDate`, `classifiedInto`, `description`)</td><td>✓</td></tr>
<tr><td>    qcabslimits                   </td><td> UNIQUE INDEX `obsInitialIdentification` (`StationId`, `ElementId`, `Datetime`)</td><td></td></tr>
<tr><td>    routinereporttransmission     </td><td> UNIQUE INDEX `report` (`reportClass`, `reportDatetime`, `reportedFrom`)</td><td></td></tr>
<tr><td>    seq_day                       </td><td> # no unique candidate: `dd` VARCHAR(50) NULL DEFAULT '0'</td><td></td></tr>
<tr><td>    seq_hour                      </td><td> # no unique candidate: `hh` INT(11) NULL DEFAULT '0'</td><td></td></tr>
<tr><td>    seq_leap_year                 </td><td> # no unique candidate: `yyyy` INT(11) NULL DEFAULT NULL</td><td></td></tr>
<tr><td>    seq_month                     </td><td> # no unique candidate: `mm` VARCHAR(50) NULL DEFAULT '0'</td><td></td></tr>
<tr><td>    seq_month_day_element         </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    seq_month_day_element_leap_yr </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    seq_year                      </td><td> # no unique candidate: `yyyy` INT(11) NULL DEFAULT NULL</td><td></td></tr>
<tr><td>    stationelement                </td><td> UNIQUE INDEX `stationElementIdentification` (`recordedFrom`, `describedBy`, `recordedWith`, `beginDate`)</td><td>✓</td></tr>
<tr><td>    stationidalias                </td><td> UNIQUE INDEX `stationid_alias_identification` (`idAlias`)</td><td></td></tr>
<tr><td>    stationlocationhistory        </td><td> UNIQUE INDEX `history` (`belongsTo`, `openingDatetime`)</td><td>✓</td></tr>
<tr><td>    stationqualifier              </td><td> UNIQUE INDEX `stationid_qualifier_identification` (`qualifier`, `qualifierBeginDate`, `qualifierEndDate`, `belongsTo`)</td><td>✓</td></tr>
<tr><td>    tdcf_indicators               </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    tm_307073                     </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    tm_307080                     </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    tm_307081                     </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    tm_307082                     </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    tm_307083                     </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    tm_307084                     </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    tm_307086                     </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
<tr><td>    tm_307089                     </td><td> # no unique candidate (table may contain duplicates)</td><td></td></tr>
</table>

**Comment by Steve-Palmer** [comment]
In the list of tables without primary keys, there are several named "aws_xxxxx" where the xxxxx looks like a country or group (Malawi, Rwanda, REMA). It looks as though these are either temporary tables created during the setup of station details or they are used in processing automatic weather station reports from the named country or group. Either way, these probably should not be in the distribution database.

**Comment by smachua** [comment]
@Steve-Palmer you're right about these tables. The "aws_xx.." tables are created at run time whenever a new AWS data structure is introduced in Climsoft operation. Those listed by @isedwards have been left as samples in the distribution database as a demonstration on how a new AWS data structure should be constructed.
There also tables for temporary use e.g. bufr_crex_data which is created to combine elements from a TDCF Template (i.e. tm_307xxxx) in use with those from bufr_crex_master to produce a BUFR or CREX message. And there many more temporary tables for different Climsoft operations not listed here.
I suggest that the Primary/Unique keys be of concern only to those tables that constitute the relational data model. I have compiled and attached them here.
[Climsoft Data Model Tables and  Keys.pdf](https://github.com/climsoft/climsoft-database/files/7075655/Climsoft.Data.Model.Tables.and.Keys.pdf)
